src/Admin/AdminHelper.php

```diff
@@ -223,7 +228,25 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
      */
     public function addNewInstance($object, FieldDescriptionInterface $fieldDescription)
     {
+        @trigger_error(sprintf(
+            'Method %s() is deprecated since sonata-project/admin-bundle 3.x. It will be removed in version 4.0.'
+            .' Use %s::addInstance() instead.',
+            __METHOD__,
+            __CLASS__
+        ), E_USER_DEPRECATED);
+
         $instance = $fieldDescription->getAssociationAdmin()->getNewInstance();
+
+        return self::addInstance($object, $fieldDescription, $instance);
+    }
+
+    /**
+     * @template T of object
```

Comment:
You should use `@phpstan-template`, there were some problem with the general `@template` annotation, as it may be used for symfony templates

Reply:
You should also move the template generic to the AdminHelper class, not the method

Reply:
> You should also move the template generic to the AdminHelper class, not the method

It's allowed for method https://arnaud.le-blanc.net/post/phpstan-generics.html

In this case, I don't think the template should be moved to the class. This is not class-related.
I can use the method with an object Foo and then use it with an object Bar.

IMHO, you add the template tag when it's related to a property of the class (with construct, add, get, ...).

I still changed `template` for `phpstan-template`.